File: transpiler/src/main/java/com/here/ivi/api/model/java/JavaMethod.java
/*
 * Copyright (C) 2017 HERE Global B.V. and its affiliate(s). All rights reserved.
 *
 * This software, including documentation, is protected by copyright controlled by
 * HERE Global B.V. All rights are reserved. Copying, including reproducing, storing,
 * adapting or translating, any or all of this material requires the prior written
 * consent of HERE Global B.V. This material also contains confidential information,
 * which may not be disclosed to others without prior written consent of HERE Global B.V.
 *
 */

package com.here.ivi.api.model.java;

import java.util.*;
import java.util.stream.Stream;
import lombok.Singular;

public final class JavaMethod extends JavaElement {

  public enum MethodQualifier {
    STATIC("static"),
    NATIVE("native");

    private final String value;

    MethodQualifier(final String value) {
      this.value = value;
    }

    @Override
    public String toString() {
      return value;
    }
  }

  public final JavaType returnType;
  public final String returnComment;
  public final Set<MethodQualifier> qualifiers;
  public final List<JavaParameter> parameters;
  public final JavaCustomType exception;

  @lombok.Builder(builderClassName = "Builder", toBuilder = true)
  private JavaMethod(
      final String name,
      final JavaVisibility visibility,
      final JavaType returnType,
      final String returnComment,
      final JavaCustomType exception,
      @Singular Set<MethodQualifier> qualifiers,
      @Singular List<JavaParameter> parameters) {
    super(name);
    this.visibility = visibility;
    this.returnType = returnType != null ? returnType : JavaPrimitiveType.VOID;
    this.returnComment = returnComment;
    this.exception = exception;
    this.qualifiers =
        qualifiers != null && !qualifiers.isEmpty()
            ? qualifiers
            : EnumSet.noneOf(MethodQualifier.class);
    this.parameters = parameters != null ? new LinkedList<>(parameters) : new LinkedList<>();
  }

  public static Builder builder(final String name) {
    return new Builder().name(name);
  }

  @Override
  public Stream<JavaElement> stream() {
    return Stream.concat(Stream.of(returnType, exception), parameters.stream());
  }
}

File: genium/src/main/java/com/here/genium/model/java/JavaMethod.kt
/*
 * Copyright (C) 2016-2019 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.genium.model.java

import com.here.genium.model.common.Comments
import java.util.EnumSet
import java.util.stream.Stream

class JavaMethod(
    name: String,
    comment: Comments = Comments(),
    visibility: JavaVisibility = JavaVisibility.PUBLIC,
    val returnType: JavaType = JavaPrimitiveType.VOID,
    val returnComment: String? = null,
    val exception: JavaCustomType? = null,
    val parameters: List<JavaParameter> = emptyList(),
    val isConstructor: Boolean = false,
    qualifiers: Set<MethodQualifier> = EnumSet.noneOf(MethodQualifier::class.java),
    annotations: Set<JavaType> = emptySet()
) : JavaElement(name) {

    val qualifiers: MutableSet<MethodQualifier> = LinkedHashSet(qualifiers)

    init {
        this.comment = comment
        this.visibility = visibility
        this.annotations.addAll(annotations)
    }

    @Suppress("unused")
    val allAnnotations
        get() = returnType.annotations + annotations

    enum class MethodQualifier constructor(private val value: String) {
        STATIC("static"),
        NATIVE("native");

        override fun toString() = value
    }

    override fun stream() = Stream.concat(
        Stream.of(returnType, exception).filter { it != null },
        Stream.concat(parameters.stream(), super.stream()))

    fun shallowCopy() = JavaMethod(
        name,
        comment,
        visibility,
        returnType,
        returnComment,
        exception,
        parameters,
        isConstructor,
        qualifiers,
        annotations
    )
}
